stream: squelch ECONNRESET error if already closed

santigimeno opened the pull request:
- Add new UV__POLLRDHUP event to be emitted when EPOLLRDHUP(in Linux) or
  EV_EOF(in BSD) is detected.
- When a read returns ECONNRESET after a UV__POLLRDHUP event, emit EOF instead
  of the error.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -1140,6 +1140,11 @@ static void uv__read(uv_stream_t* stream) {
         }
         stream->read_cb(stream, 0, &buf);
       } else {
+        if ((errno == ECONNRESET) && (stream->flags & UV_STREAM_DISCONNECT)) {
```
The parens aren't necessary around the first clause.

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -1140,6 +1140,11 @@ static void uv__read(uv_stream_t* stream) {
       } else {
+        if ((errno == ECONNRESET) && (stream->flags & UV_STREAM_DISCONNECT)) {
+          uv__stream_eof(stream, &buf);
+          return;
```
Seeing how you have a return statement here, you might as well fold the if statement into the if/else statement above it (and then you don't need the return statement either because there is one below.)

bnoordhuis commented on src/unix/stream.c:
```diff
@@ -1227,8 +1232,13 @@ static void uv__stream_io(uv_loop_t* loop, uv__io_t* w, unsigned int events) {
+  if (events & (UV__POLLIN | UV__POLLERR | UV__POLLHUP | UV__POLLRDHUP)) {
+    if (events & UV__POLLRDHUP) {
+      stream->flags |= UV_STREAM_DISCONNECT;
+    }
```
Can you drop the braces here?

bnoordhuis commented:
CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/110/

santigimeno commented:
@bnoordhuis updated following your comments.

test-poll.c are failing in OSX and FreeBSD's. I'll look into them

santigimeno commented:
PR updated with a couple of changes. Now the tests look ok but I'm not sure if the changes in `test-poll.c` make sense
/cc @bnoordhuis

saghul commented:
I don't know why those changes to test-poll.c are necessary. The poll callback will be called when the socket is readable, so how come we now get EAGAIN on the test?

santigimeno commented:
@saghul yes, I don't really understand what's going on either. Let me try to explain what I was seeing.

First of all, I am seeing this issue only in FreeBSD (from the CI output, in OSX too) and happens both in the `poll_unidirectional` and `poll_duplex` tests

After the initial PR, the problem was the assertion in https://github.com/libuv/libuv/blob/v1.x/test/test-poll.c#L193, where `events = 0` and `context->events = 1`, thus the change I made here: https://github.com/libuv/libuv/pull/403/files#diff-ea2913f17cbe1466e25e48a517e89771R44, to emit `UV_READABLE` when `UV__POLLRDHUP` is received.

After this change the assertion moved down to https://github.com/libuv/libuv/blob/v1.x/test/test-poll.c#L225 due to the `EAGAIN` error when reading.

When analyzing what's happening, in the context of a `poll_unidirectional` test, I see the following sequence of events related to the socket that *fails*:
- It connects to the server.
- It receives an `UV_WRITABLE` event and sends `TRANSFER_BYTES` data until `EAGAIN`
- It receives an `UV_WRITABLE` event and calls `shutdown`.
- It receives an `UV_READABLE` due to an `UV__POLLRDHUP` due to an `EV_EOF` and the `recv` returns `EAGAIN`

In the meantime the server side has only accepted the connection.

I hope it makes some sense to you. Thanks!

saghul commented:
@santigimeno Thanks for the thorough explanation. I see.

Since this touches a quite sensitive part of libuv I'd like to see a test which fails before this patch. You can go platform-dependent and ifdef the test out on Windows or other platforms.

I have the following test failure os OSX 10.9:

~~~~
`tcp_open` failed: exit code 134
Output from process `tcp_open`: (no output)
Output from process `tcp_open`:
Assertion failed in ../test/test-tcp-open.c on line 114: nread == 4
~~~~

Last, how about the other I/O backends (Solaris, AIX)?

santigimeno commented:
@saghul I have added a test that fails before this patch in `FreeBSD` but passes after.

About that test, I don't have access to a OSX 10.9 machine but can try in a OSX 10.10

About the other backends, I don't know if they have equivalents to `EV_EOF` or `EPOLLRDHUP`

santigimeno commented:
@saghul I have updated the PR with a fix for `test-tcp-open`: in OS X it was getting `EAGAIN` in the first `read` call. Also, I have rebased to current `v1.x`

santigimeno commented:
Any additional feedback on this? Thanks! :)

bnoordhuis commented on test/test-tcp-open.c:
```diff
@@ -111,11 +112,15 @@ static void read_cb(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf) {
+    read_bytes += nread;
+    if (nread > 0) {
+        ASSERT(nread == 4);
+        ASSERT(memcmp("PING", buf->base, nread) == 0);
```
Style nit: can you use two spaces for indent here?

bnoordhuis commented on test/test-tcp-squelch-connreset.c:
```diff
@@ -0,0 +1,119 @@
+  r = uv_write(&write_req, (uv_stream_t*)(req->handle), &outbuf, 1, NULL);
+  ASSERT(r == 0);
+
+  r = uv_read_start((uv_stream_t*)(req->handle), alloc_cb, read_cb);
```
Style nit: the parens around `req->handle` are not strictly necessary and a little incongruent with the style in the rest of the file.

bnoordhuis commented on test/test-tcp-squelch-connreset.c:
```diff
@@ -0,0 +1,119 @@
+                     &tcp_client,
+                     (const struct sockaddr*) &addr,
+                     on_connect);
+  ASSERT(r == 0);;
```
Style nit: a semi-colon too many.

bnoordhuis commented:
A couple of style nits but basically LGTM.  CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/116/ (lots of red expected, unfortunately.)

saghul commented:
I have a general question: since this is only implemented for epoll and kqueue, Node and other projects still need to expect ECONNRESET, right? So, what do we gain in the end? (sorry if I have missed something along the way)

bnoordhuis commented:
Applications will still need to handle ECONNRESET because it can happen legitimately but it should reduce the number of "frivolous" ECONNRESET events on the main UNIX platforms, Linux and OS X.

Aside, I think it's possible to make Solaris and AIX behave the same way but they're fringe enough that I don't care if it's not implemented right away.

saghul commented:
Test fails on Windows:

~~~~
`tcp_squelch_connreset` failed: exit code 3
Output from process `tcp_squelch_connreset`:
Assertion failed in test\test-tcp-squelch-connreset.c on line 43: nread == UV_EOF
~~~~

santigimeno commented:
@saghul the changes in the patch are unrelated to Windows and it's quite a simple test... I've no access to a Windows box, would it be possible to launch a new CI execution to get the value of `nread`?

saghul commented:
@santigimeno you can push debug commits printing stuff and @bnoordhuis or I can start new CI jobs for you.

santigimeno commented:
I have fixed style issues and added a debug log to see what's going on on Windows.
@bnoordhuis @saghul could you please start a new CI? Thanks

saghul commented:
@santigimeno https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/117/

santigimeno commented:
It returns `ECONNRESET` too. I don't know if there's a way to squelch it too like in the unixes or just run the test in the UNIX boxes.

saghul commented:
Since this patch only avoids the error on Linux and BSDs, I guess it should only run on those.

Generally speaking, this makes me a bit uneasy, so I'm -0. But @bnoordhuis approved, so let's go ahead.

santigimeno commented:
I didn't expect the `ECONNRESET` issue to happen in Windows too. I would be more comfortable if the Windows case could be covered too, but don't have the knowledge.

santigimeno commented:
Rebased to current v1.x and squashed into a single commit.
`tcp_squelch_connreset` disabled in Windows (I think I've done it right)

saghul commented:
CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/120/

saghul commented:
Thanks Santi, landed in 05a003a!

santigimeno commented:
Thanks!
